detectable/src/main/java/com/synopsys/integration/detectable/detectables/setuptools/transform/SetupToolsGraphTransformer.java

```diff
@@ -0,0 +1,152 @@
+package com.synopsys.integration.detectable.detectables.setuptools.transform;
+
+import java.io.File;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import org.apache.commons.lang3.StringUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.synopsys.integration.bdio.graph.BasicDependencyGraph;
+import com.synopsys.integration.bdio.graph.DependencyGraph;
+import com.synopsys.integration.bdio.model.Forge;
+import com.synopsys.integration.bdio.model.dependency.Dependency;
+import com.synopsys.integration.bdio.model.externalid.ExternalId;
+import com.synopsys.integration.bdio.model.externalid.ExternalIdFactory;
+import com.synopsys.integration.detectable.ExecutableTarget;
+import com.synopsys.integration.detectable.ExecutableUtils;
+import com.synopsys.integration.detectable.detectable.executable.DetectableExecutableRunner;
+import com.synopsys.integration.detectable.detectables.setuptools.parse.SetupToolsParsedResult;
+import com.synopsys.integration.detectable.python.util.PythonDependency;
+import com.synopsys.integration.executable.ExecutableOutput;
+import com.synopsys.integration.executable.ExecutableRunnerException;
+
+public class SetupToolsGraphTransformer {
+
+    private final Logger logger = LoggerFactory.getLogger(this.getClass());
+
+    private static final String REQUIRES = "Requires";
+    private static final String UNEXPECTED_PIP_OUTPUT = "Unexpected empty pip show results. Please run pip install . on the project and try running Detect again.";
+    
+    private File sourceDirectory;
+    private ExternalIdFactory externalIdFactory;
+    private DetectableExecutableRunner executableRunner;
+
+    public SetupToolsGraphTransformer(File sourceDirectory, ExternalIdFactory externalIdFactory, DetectableExecutableRunner executableRunner) {
+        this.sourceDirectory = sourceDirectory;
+        this.externalIdFactory = externalIdFactory;
+        this.executableRunner = executableRunner;
+    }
+
+    public DependencyGraph transform(ExecutableTarget pipExe, SetupToolsParsedResult parsedResult) throws ExecutableRunnerException {
+        DependencyGraph dependencyGraph = new BasicDependencyGraph();
+        
+        if (pipExe != null) {
+            // Get dependencies by running pip show on each direct dependency
+            for (PythonDependency directDependency : parsedResult.getDirectDependencies()) {
+                handleShowDependency(pipExe, dependencyGraph, directDependency.getName(), null, directDependency.isConditional());
+            }
+        } else {
+            // Unable to determine transitive dependencies, add parsed dependencies directly
+            // to the root of the graph.
+            handleParsedDependencies(parsedResult, dependencyGraph);
+        }
```

Comment:
Can pipExe be null at this point? Won't the extractable method return False in the upstream Detectable code if pipExe is null? Curious when this method would run - was it the idea to use it as a fallback if the pip show command in handleShowDependency() fails?

Reply:
This class is used by both the build and the buildless detectable. For the build detectable you are right, it will never get here. It will fail cause it can't find pip during the extractable method. But for the buildless it won't care about pip so the else is for the buildless detectable. I actually did think about falling back to this in the build detector at one point but ultimately just said let the cascade call the buildless detector to handle it.

Reply:
Got it, then this does make sense in the buildless flow so I'm good with it. Thanks!